hamilton/plugins/dlt_extensions.py

```diff
@@ -0,0 +1,74 @@
+import dataclasses
+from typing import Any, Collection, Dict, Iterable, Literal, Optional, Type
+
+import dlt
+import pandas as pd
+
+# importing TDestinationReferenceArg fails if Destination isn't imported
+from dlt.common.destination import Destination, TDestinationReferenceArg  # noqa: F401
+
+from hamilton import registry
+from hamilton.io.data_adapters import DataSaver
+
+try:
+    import pyarrow as pa
+except ModuleNotFoundError:
+    SAVER_TYPES = [Iterable, pd.DataFrame]
+else:
+    SAVER_TYPES = [Iterable, pd.DataFrame, pa.Table, pa.RecordBatch]
+
+
+@dataclasses.dataclass
+class DltSaver(DataSaver):
+    """Materialize results using a dlt pipeline with the specified destination.
+
+    In reference to an Extract, Transform, Load (ETL) pipeline, here, the Hamilton
+    dataflow is responsible for Transform, and `DltDestination` for Load.
+    """
+
+    pipeline: dlt.Pipeline
+    rename: Optional[dict] = None
+    # kwargs for pipeline.run()
+    write_disposition: Optional[Literal["skip", "append", "replace", "merge"]] = None
+
+    @classmethod
+    def applicable_types(cls) -> Collection[Type]:
+        return SAVER_TYPES
+
+    def save_data(self, data) -> Dict[str, Any]:
+        """
+        ref:
+            tables: https://dlthub.com/docs/dlt-ecosystem/verified-sources/arrow-pandas
+            standalone resources: https://dlthub.com/docs/general-usage/resource#load-resources
+            dynamic resources: https://dlthub.com/docs/general-usage/source#create-resources-dynamically
+        """
+        if isinstance(data, dict) is False:  # only 1 node, no `combine=base.DictResult()`
+            raise NotImplementedError(
+                f"DltSaver needs to a receive a `Dict[str, {SAVER_TYPES}]`"
+                f"When using `to.dlt()`, make sure to specify `combine=base.DictResult()`"
+            )
+
+        rename = {} if self.rename is None else self.rename
+
+        resources = []
+        for node_name, result in data.items():
+            name = rename.get(node_name, node_name)
+            resources.append(dlt.resource(result, name=name))
+
+        load_info = self.pipeline.run(resources, write_disposition=self.write_disposition)
+        return load_info.asdict()
```

Comment:
Let's put this behind `dlt_data` or some field so we can add our own metadata as well

Reply:
what metadata should we return given we don't have a file path to check for

Reply:
Meh no idea, but it fits in with the rest of the pattern if we have it namespaced. E.G. this: https://github.com/DAGWorks-Inc/hamilton/blob/5dd6ab4bf8933aeb71b78e1c28045afc7e46623c/hamilton/io/utils.py#L42

Reply:
We should return metadata that could help us identify/understand this artifact better. It might be the same that's in the dict? or some subset of that?